### PR TITLE
Caching hook on WistiaPlayerDelegate

### PR DESCRIPTION
**Pull request by hardythedrummer**

### CODE REVIEW

**Comment by hardythedrummer** [comment]
Not sure whats up with tests - they all pass locally...

File: Pod/Classes/Playback/internal/_WistiaPlayer.swift [hunk 1]
```diff
@@ -92,6 +92,10 @@ internal extension WistiaPlayer {
     //
     //We just need HLS (if required), otherwise mp4.  If there are options, we pick the best sized.
     internal func bestPlaybackUrl(for media: WistiaMedia, andAssetWithSlug assetSlug: String?, requiringHLS requireHLS: Bool, atTargetWidth targetWidth: Int64) throws -> URL {
+        let delegateUrl = delegate?.wistiaPlayer(self, shouldLoadVideoForMedia: media)
+        guard delegateUrl == nil else {
```
Comment (spinosa) [inline Pod/Classes/Playback/internal/_WistiaPlayer.swift]: let's simplify this by combining the two above lines into a single `if let`

File: Pod/Classes/Playback/internal/_WistiaPlayer.swift [hunk 2]
```diff
@@ -92,6 +92,10 @@ internal extension WistiaPlayer {
...
+            return delegateUrl!
```
Comment (spinosa) [inline Pod/Classes/Playback/internal/_WistiaPlayer.swift]: need to call `delegate?.wistiaPlayer(self, willLoadVideoForMedia: media, usingAsset: nil, usingHLSMasterIndexManifest: false)` before returning

File: Pod/Classes/Playback/public/WistiaPlayer.swift [hunk 3]
```diff
@@ -89,6 +89,14 @@ public protocol WistiaPlayerDelegate : class {
 
      */
     func wistiaPlayer(_ player:WistiaPlayer, willLoadVideoForMedia media:WistiaMedia, usingAsset asset:WistiaAsset?, usingHLSMasterIndexManifest: Bool)
+    
+    /**
+     Informs the delegate that the WistiaPlayer is attempting to determine a URL for loading a `WistiaAsset`, giving an opportunity
+     for the delegate to inject a url (such as a cached local asset url).
+
```
Comment (spinosa) [inline Pod/Classes/Playback/public/WistiaPlayer.swift]: Add 
`- Parameter player: The `WistiaPlayer` determining the URL for playback.`

File: Pod/Classes/Playback/public/WistiaPlayer.swift [hunk 4]
```diff
@@ -89,6 +89,14 @@ public protocol WistiaPlayerDelegate : class {
 
      */
     func wistiaPlayer(_ player:WistiaPlayer, willLoadVideoForMedia media:WistiaMedia, usingAsset asset:WistiaAsset?, usingHLSMasterIndexManifest: Bool)
+    
+    /**
+     Informs the delegate that the WistiaPlayer is attempting to determine a URL for loading a `WistiaAsset`, giving an opportunity
+     for the delegate to inject a url (such as a cached local asset url).
+     
+     - Parameter media: The 'WistiaMedia' from which an asset will be chosen.
```
Comment (spinosa) [inline Pod/Classes/Playback/public/WistiaPlayer.swift]: This should probably read something like

`- Parameter media: The 'WistiaMedia' for which a playback URL is needed.`

File: Pod/Classes/Playback/public/WistiaPlayer.swift (hunk 3 above)
Comment (spinosa) [inline Pod/Classes/Playback/public/WistiaPlayer.swift]: Let's add some more descriptive information here

`The `WistiaPlayer` will use the `URL` returned by this method for playback.  Return `nil` to allow the `WistiaPlayer` to determine the best playback URL.`

**Comment by hardythedrummer** [comment]
Addressed your comments, please view again @spinosa

File: Pod/Classes/Playback/public/WistiaPlayer.swift [hunk 5]
```diff
@@ -82,13 +82,24 @@ public protocol WistiaPlayerDelegate : class {
...
+    /**
+     Informs the delegate that the `WistiaPlayer` is attempting to determine a URL for loading a `WistiaAsset`, giving an opportunity
+     for the delegate to inject a url (such as a cached local asset url). The `WistiaPlayer` will use the URL returned by this method
+     for playback. Return nil to allow the `WistiaPlayer` to determine the best playback URL.
+     
+     - Parameter player: The `WistiaPlayer` determining the URL for playback.
+     - Parameter media: The 'WistiaMedia' from which a playback URL is needed.
+     */
+    func wistiaPlayer(_ player:WistiaPlayer, shouldLoadVideoForMedia media:WistiaMedia) -> URL?
 }
```
Comment (spinosa) [inline Pod/Classes/Playback/public/WistiaPlayer.swift]: If we add the following protocol extension here, current users of this protocol (including `WistiaPlayerViewController` within WK) wouldn't need to make any changes and wouldn't be affected.

```
public extension WistiaPlayerDelegate {

    //Default implementation retains standard functionality
    public func wistiaPlayer(_ player: WistiaPlayer, shouldLoadVideoForMedia media: WistiaMedia) -> URL? {
        return nil
    }

}
```

File: Pod/Classes/Playback/public/WistiaPlayer.swift [hunk 6]
```diff
@@ -82,13 +82,24 @@ public protocol WistiaPlayerDelegate : class {
...
         (aka Manifest of Manifests). `asset` will be `nil`.
 
      */
     func wistiaPlayer(_ player:WistiaPlayer, willLoadVideoForMedia media:WistiaMedia, usingAsset asset:WistiaAsset?, usingHLSMasterIndexManifest: Bool)
+    
+    /**
+     Informs the delegate that the `WistiaPlayer` is attempting to determine a URL for loading a `WistiaAsset`, giving an opportunity
+     for the delegate to inject a url (such as a cached local asset url). The `WistiaPlayer` will use the URL returned by this method
+     for playback. Return nil to allow the `WistiaPlayer` to determine the best playback URL.
+
```
Comment (spinosa) [inline Pod/Classes/Playback/public/WistiaPlayer.swift]: We should also add a note that this is an optional method:

> Optional.  The default implementation returns `nil`, retaining standard functionality.

File: Pod/Classes/Playback/internal/view controller/_WistiaPlayerViewController.swift [hunk 7]
```diff
@@ -279,7 +279,11 @@ extension WistiaPlayerViewController: WistiaPlayerDelegate {
     public final func wistiaPlayer(_ player: WistiaPlayer, willLoadVideoForMedia media: WistiaMedia, usingAsset asset: WistiaAsset?, usingHLSMasterIndexManifest: Bool) {
         configurePlayerViews(for: media)
     }
-
+    
+    /// Internal
+    public final func wistiaPlayer(_ player: WistiaPlayer, shouldLoadVideoForMedia media: WistiaMedia) -> URL? {
```
Comment (spinosa) [inline Pod/Classes/Playback/internal/view controller/_WistiaPlayerViewController.swift]: We can get rid of this entirely.  It's opt-in, no code changes required!

**Comment by hardythedrummer** [comment]
Sounds good to me @spinosa . Changes made, please review again.

**Pull request merged**